### ISSUE TITLE
Duplicate identifier when using `identifierNamesGenerator: 'mangled'`

### ISSUE THREAD

**Issue by slig** [javascript-obfuscator/javascript-obfuscator#549]
Hi!

I'm not sure if this constitutes a bug or not, as I haven't seen a proper error yet, only a warning from `SizeLimitsPlugin`:

`Error parsing bundle asset "/workspace/frontend/dist/js/app-bundle-9144a3e50714133abf44.js": Identifier 'k' has already been declared (1:3083)`

## Expected Behavior
The `identifierNamesGenerator` should produce unique identifiers within the same scope.

## Current Behavior

Notice two `k` identifiers bellow:

```
    const k = function() {
        let H = !![];
        return function(I, J) {
            const K = H ? function() {
                if (J) {
                    const L = J[b('0x56')](I, arguments);
                    J = null;
                    return L;
                }
            } : function() {};
            H = ![];
            return K;
        };
    }();

    function k(M) {
        const hk = f(this, function() {
        .....
```

## Steps to Reproduce
I'm using `webpack-ofuscator@0.24.6` with the following options:

```
const OBFUSCATOR_OPTIONS = {
  seed: 1984,
  selfDefending: true,
  identifierNamesGenerator: 'mangled',
  domainLock: ['example.local', 'example.com'],
};
```

## Your Environment
* Obfuscator version used: 0.24.6
* Node version used: 10.17.0

Thanks for the all the hard work on this project!

**Comment by sanex3339** [comment]
Hi, can you show this two functions before obfuscation?

**Comment by slig** [comment]
Hi,

It looks like the code is part of the `domainLock` feature, here's what the second `k` looks like:


```
    function k(M) {
        const hk = f(this, function() {
            const c = function() {
                    return '\x64\x65\x76';
                },
                d = function() {
                    return '\x77\x69\x6e\x64\x6f\x77';
                };
            const e = function() {
                const i = new RegExp('\x5c\x77\x2b\x20\x2a\x5c\x28\x5c\x29\x20\x2a\x7b\x5c\x77\x2b\x20\x2a\x5b\x27\x7c\x22\x5d\x2e\x2b\x5b\x27\x7c\x22\x5d\x3b\x3f\x20\x2a\x7d');
                return !i['\x74\x65\x73\x74'](c['\x74\x6f\x53\x74\x72\x69\x6e\x67']());
            };
            const f = function() {
                const j = new RegExp('\x28\x5c\x5c\x5b\x78\x7c\x75\x5d\x28\x5c\x77\x29\x7b\x32\x2c\x34\x7d\x29\x2b');
                return j['\x74\x65\x73\x74'](d['\x74\x6f\x53\x74\x72\x69\x6e\x67']());
            };
            const g = function(k) {
                const l = ~-0x1 >> 0x1 + 0xff % 0x0;
                if (k['\x69\x6e\x64\x65\x78\x4f\x66']('\x69' === l)) {
                    h(k);
                }
            };
            const h = function(m) {
                const n = ~-0x4 >> 0x1 + 0xff % 0x0;
                if (m['\x69\x6e\x64\x65\x78\x4f\x66']((!![] + '')[0x3]) !== n) {
                    g(m);
                }
            };
            if (!e()) {
                if (!f()) {
                    g('\x69\x6e\x64\u0435\x78\x4f\x66');
                } else {
                    g('\x69\x6e\x64\x65\x78\x4f\x66');
                }
            } else {
                g('\x69\x6e\x64\u0435\x78\x4f\x66');
            }
        });
        hk();
   ....
```

**Comment by sanex3339** [comment]
Dont understand how you get two same identifier names for two helper functions (one for domain lock, one for self-defending).
Cannot reproduce, for me this identifiers are always different

```
var k = function() {
        var c = !![];
        return function(d, e) {
            var f = c ? function() {
                if (e) {
                    var g = e['apply'](d, arguments);
                    e = null;
                    return g;
                }
            } : function() {};
            c = ![];
            return f;
        };
    }();
    var a0 = k(this, function() {
        var c = function() {
                return '\x64\x65\x76';
            },
            d = function() {
                return '\x77\x69\x6e\x64\x6f\x77';
            };
        var e = function() {
            var i = new RegExp('\x5c\x77\x2b\x20\x2a\x5c\x28\x5c\x29\x20\x2a\x7b\x5c\x77\x2b\x20\x2a\x5b\x27\x7c\x22\x5d\x2e\x2b\x5b\x27\x7c\x22\x5d\x3b\x3f\x20\x2a\x7d');
            return !i['\x74\x65\x73\x74'](c['\x74\x6f\x53\x74\x72\x69\x6e\x67']());
        };
        var f = function() {
            var j = new RegExp('\x28\x5c\x5c\x5b\x78\x7c\x75\x5d\x28\x5c\x77\x29\x7b\x32\x2c\x34\x7d\x29\x2b');
            return j['\x74\x65\x73\x74'](d['\x74\x6f\x53\x74\x72\x69\x6e\x67']());
        };
        var g = function(k) {
            var l = ~-0x1 >> 0x1 + 0xff % 0x0;
            if (k['\x69\x6e\x64\x65\x78\x4f\x66']('\x69' === l)) {
                h(k);
            }
        };
        var h = function(m) {
            var n = ~-0x4 >> 0x1 + 0xff % 0x0;
            if (m['\x69\x6e\x64\x65\x78\x4f\x66']((!![] + '')[0x3]) !== n) {
                g(m);
            }
        };
        if (!e()) {
            if (!f()) {
                g('\x69\x6e\x64\u0435\x78\x4f\x66');
            } else {
                g('\x69\x6e\x64\x65\x78\x4f\x66');
            }
        } else {
            g('\x69\x6e\x64\u0435\x78\x4f\x66');
        }
    });
```

**Comment by slig** [comment]
Thanks for looking into this! I'll extract this specific file and try to reproduce independently of the current setup (webpack).

**Comment by sanex3339** [comment]
This issue is probably has the same problem as  https://github.com/javascript-obfuscator/javascript-obfuscator/issues/550

So, i want to improve variable names preserving.

**Comment by slig** [comment]
I originally had `transformObjectKeys` enabled and had issues (the code din't run) as well, so I disabled and it runs, but got this duplicated identifier.

**Comment by sanex3339** [comment]
The options itself do not affect this behaviour. The problem is more complex: a generated identifier can be the same as an existing identifier in the code. So, i can fix it in two ways:
- fast and stupid. Just remember all identifiers in the source code and regenerate identifier name if it the same as the one in the stored names.
- more complex - do the same thing, but stored identifier names are attached to their lexical scopes.

Currently, i'm trying to implement the second approach

**Comment by sanex3339** [comment]
Can you try this branch?
https://github.com/javascript-obfuscator/javascript-obfuscator/pull/552

Will it help?

**Comment by slig** [comment]
Can confirm that `identifierNamesGenerator` doesn't produce duplicated identifiers anymore, but `transformObjectKeys` doesn't works yet. Here's the runtime error I get:

`Uncaught ReferenceError: Must call super constructor in derived class before accessing 'this' or returning from derived constructor`

Here's the obfuscated code snippet:

```
        var d0 = cZ(0x0)
          , d1 = cZ['n'](d0)
          , d2 = cZ(0x26)
          , d3 = cZ(0x25);
        class d4 extends d1['a'][a4('0x60')] {
            constructor(d7) {
                var d8 = {};
                d8[a4('0x40')] = this['props'][a4('0x19')]; // <<<< line with error
                super(d7),
                this[a4('0x113')] = d8,
                this['close'] = this['close'][a4('0x15')](this);
            }

```

And here's the original TS code:

```
class ModalEndGame extends React.Component<Props, State> {

  constructor(props: Props) {
    super(props);

    this.state = {
      open: this.props.ended,
    };

    this.close = this.close.bind(this);
  }
 .....
```

Thank you very much for looking into this!

**Comment by sanex3339** [comment]
@slig try now

**Comment by slig** [comment]
I'm having the same issues as before. I deleted the `node_modules/javascript-obfuscator` and `node_modules/webpack-obfuscator/node_modules/javascript-obfuscator` and installed again using:

`yarn add https://github.com/javascript-obfuscator/javascript-obfuscator.git#variable-preserve-improvements`

to make sure.

**Comment by sanex3339** [comment]
That is strange. My tests was fixed.

**Comment by slig** [comment]
I must be making some stupid mistake. Will try more carefully later today / earlier tomorrow. Thanks again!

**Comment by sanex3339** [comment]
What version number did you see, when obfuscated you code? `0.25.0`?

**Comment by slig** [comment]
Hi, yes, `0.25.0`. This time though it seems like `transforObjectKeys` is working. 

After using:

`yarn add https://github.com/javascript-obfuscator/javascript-obfuscator.git#variable-preserve-improvements`

I made sure there was not an `javascript-obfuscator` inside `webpack-obfuscator/node_modules` and checked to see if this new file was present: `cat node_modules/javascript-obfuscator/test/functional-tests/generators/identifier-names-generators/mangled-identifier-names-generator/fixtures/string-array-storage-name-conflict-2.js`.

I've sent you an email with the code that produces this duplicated error.

Thanks again!

**Comment by sanex3339** [comment]
Can you try this branch? 
https://github.com/javascript-obfuscator/javascript-obfuscator/pull/555/

**Comment by slig** [comment]
Hi, still getting duplicated identifiers at runtime. Thanks again!

**Comment by sanex3339** [comment]
Did you get 
```
Uncaught SyntaxError: Identifier 'q' has already been declared?
```

**Comment by sanex3339** [comment]
Reproduced it. Will check

**Comment by sanex3339** [comment]
Can be easily reproduced on `obfuscator.io` with:
```
const c = 1;
```

and

```
identifierNamesGenerator: 'mangled';
domainLock: true;
```
All other options should be disabled
All options

**Comment by sanex3339** [comment]
Updated PR. Please try now.

**Comment by slig** [comment]
Sorry, somehow missed this notification. Will try later today. Thanks again!

**Comment by sanex3339** [comment]
Now all tests should be on `variable-preserve-improvements` branch

**Comment by sanex3339** [comment]
Released as `0.25.0`
